src/vs/workbench/services/remote/node/tunnelService.ts

```diff
@@ -116,33 +125,51 @@ export class TunnelService implements ITunnelService {
 		this._tunnels.clear();
 	}
 
-	openTunnel(remotePort: number): Promise<RemoteTunnel> | undefined {
+	openTunnel(remotePort: number, localPort: number): Promise<RemoteTunnel> | undefined {
 		const remoteAuthority = this.environmentService.configuration.remoteAuthority;
 		if (!remoteAuthority) {
 			return undefined;
 		}
 
-		const resolvedTunnel = this.retainOrCreateTunnel(remoteAuthority, remotePort);
+		const resolvedTunnel = this.retainOrCreateTunnel(remoteAuthority, remotePort, localPort);
 		if (!resolvedTunnel) {
 			return resolvedTunnel;
 		}
 
-		return resolvedTunnel.then(tunnel => ({
+		return resolvedTunnel.then(tunnel => {
+			const newTunnel = this.makeTunnel(tunnel);
+			this._onTunnelOpened.fire(newTunnel);
+			return newTunnel;
+		});
+	}
+
+	private makeTunnel(tunnel: RemoteTunnel): RemoteTunnel {
+		return {
 			tunnelRemotePort: tunnel.tunnelRemotePort,
 			tunnelLocalPort: tunnel.tunnelLocalPort,
+			localAddress: tunnel.localAddress,
 			dispose: () => {
-				const existing = this._tunnels.get(remotePort);
+				const existing = this._tunnels.get(tunnel.tunnelRemotePort);
 				if (existing) {
 					if (--existing.refcount <= 0) {
 						existing.value.then(tunnel => tunnel.dispose());
-						this._tunnels.delete(remotePort);
+						this._tunnels.delete(tunnel.tunnelRemotePort);
+						this._onTunnelClosed.fire(tunnel.tunnelRemotePort);
 					}
 				}
 			}
-		}));
+		};
+	}
+
+	async closeTunnel(remotePort: number): Promise<void> {
+		if (this._tunnels.has(remotePort)) {
+			const value = this._tunnels.get(remotePort)!;
+			value.refcount = 0;
+			(await value.value).dispose();
```

Comment:
The await should happen before recount is set to zero. Otherwise I think there's a possible race where someone grabs the tunnel between between the await and the call to dispose

Reply:
Good point, thanks!